packages/SwingSet/src/kernel/kernel.js

```diff
@@ -28,16 +28,124 @@ function abbreviateReviver(_, arg) {
+  // more (encapsulate importBundle?) and provide a single sensible thing to
+  // vats.
+  let complainedAboutGlobalMetering = false;
+  const allRefillers = new Set(); // refill() functions
+  function makeGetMeter(options = {}) {
+    const { refillEachCrank = true, refillIfExhausted = true } = options;
+
+    if (!replaceGlobalMeter && !complainedAboutGlobalMetering) {
+      console.error(
+        `note: makeGetMeter() cannot enable global metering, app must import @agoric/install-metering-and-ses`,
+      );
+      // to fix this, your application program must
+      // `import('@agoric/install-metering-and-ses')` instead of
+      // `import('@agoric/install-ses')`
+      complainedAboutGlobalMetering = true;
+    }
+
+    // zoe's importBundle(dapp-encouragement contract) causes babel to use
+    // about 5.4M computrons and 6.4M allocatrons (total 11.8M) in a single
+    // crank, so the default of 1e7 is insufficient
+    const FULL = 1e8;
+    const { meter, refillFacet } = makeMeter({
+      budgetAllocate: FULL,
+      budgetCompute: FULL,
+    });
+    function refill() {
+      // console.error(`-- METER REFILL`);
+      // console.error(`   allocate used: ${FULL - refillFacet.getAllocateBalance()}`);
+      // console.error(`   compute used : ${FULL - refillFacet.getComputeBalance()}`);
+      if (meter.isExhausted() && !refillIfExhausted) {
+        return;
+      }
+      refillFacet.allocate();
+      refillFacet.compute();
+      refillFacet.stack();
+    }
+    if (refillEachCrank) {
+      allRefillers.add(refill);
+    }
+    let meterUsed = false; // mostly for testing
+    function getMeter() {
+      meterUsed = true;
+      if (replaceGlobalMeter) {
+        replaceGlobalMeter(meter);
+      }
+      return meter;
+    }
+    function resetMeterUsed() {
+      meterUsed = false;
+    }
+    function getMeterUsed() {
+      return meterUsed;
+    }
+    function isExhausted() {
+      return meter.isExhausted();
+    }
+
+    // TODO: this will evolve. For now, we let the caller refill the meter as
+    // much as they want, although only tests will do this (normal vats will
+    // just rely on refillEachCrank). As this matures, vats will only be able
+    // to refill a meter by transferring credits from some other meter, so
+    // their total usage limit remains unchanged.
+    return harden({
+      getMeter,
+      isExhausted,
+      resetMeterUsed,
+      getMeterUsed,
+      refillFacet,
+      getAllocateBalance: refillFacet.getAllocateBalance,
+      getComputeBalance: refillFacet.getComputeBalance,
+      getCombinedBalance: refillFacet.getCombinedBalance,
+    });
+  }
+  harden(makeGetMeter);
+
+  function endOfCrankMeterTask() {
+    if (replaceGlobalMeter) {
+      replaceGlobalMeter(null);
+    }
+    for (const refiller of allRefillers) {
+      refiller();
+    }
+  }
+  harden(endOfCrankMeterTask);
+
+  // TODO: this doesn't work yet, it doesn't appear to prevent meter
```

Comment:
When you implement `getMeter(true)`, this TODO will be done.

Reply:
Looks like it works now, the "compute" usage during the `install` step of the agoric-cli integration test (which applies the metering transform to the dapp-encouragement contract) dropped by a factor of 10, and the "allocate" usage by a factor of 100.